Work branch: Logger expanded

erastian opened the pull request:
- HTTPError renamed to Exception
- Logger functionality expanded
- Added logger level, default = 3
- Added exceptions rules in exception filter
- added forgot-password.dto
- reworked recovery access flow with related methods
- added DB errors filters in exception.filter
- added successful log message for mailer.service
- added roles guard.middleware for comments controller
- used className for identifying methods in logs
- password hash line was excluded from method findAllUsers

erastian committed:
Work branch: Logger expanded

- HTTPError renamed to Exception
- Logger functionality expanded
- Added logger level, default = 3
- Added exceptions rules in exception filter

WIP

mdziarkach commented on src/common/guard.middleware.ts:
```diff
@@ -1,20 +1,21 @@
 	execute(req: Request, res: Response, next: NextFunction): void {
 		if (!req.user) {
-			return next(new HTTPError(StatusCodes.FORBIDDEN, 'User not authorized'));
+			return next(new Exception(StatusCodes.FORBIDDEN, 'User not authorized', this.context));
```
return next(new Exception(StatusCodes.FORBIDDEN, 'User not authorized', GuardMiddleware.name));

erastian replied:
++

mdziarkach commented on src/services/errors/exception.filter.ts:
```diff
@@ -1,22 +1,48 @@
+		if (err instanceof Exception) {
+			this.logger.error(`[${err.context}] ${err.statusCode} ${err.name}: ${err.message}.`);
 			res.status(err.statusCode).send({ err: err.message });
+		} else if (err instanceof JsonWebTokenError || err instanceof TokenExpiredError) {
```
lets remove it

erastian replied:
We still need to proceed JsonWebTokenError because without covering it passed through to 500 internal server error. But tokens expiring errors proceed well without covering in exception filter.

TokenExpiredError removed

mdziarkach commented on src/services/errors/exception.filter.ts:
```diff
@@ -1,22 +1,48 @@
+						.send({ err: `There is a unique constraint violation, a new article cannot be created with this url` });
+					break;
+				default:
+					this.logger.error(`[Database] ${err.name}: ${err.code}`);
```
lets add stack trace to log

erastian replied:
added

mdziarkach commented on src/users/users.controller.ts:
```diff
@@ -13,25 +13,30 @@ import { GuardMiddleware } from '../common/guard.middleware';
 
 @injectable()
 export class UsersController extends BaseController implements IUsersController {
+	private context = 'users';
```
let's remove it. Use please class name

erastian replied:
I will keep this parameter for transparent logging router works

erastian committed:
Work branch: Logger expanded

- added forgot-password.dto
- reworked recovery access flow with related methods
- added DB errors filters in exception.filter
- added successful log message for mailer.service
- added roles guard.middleware for comments controller
- used className for identifying methods in logs
- password hash line was excluded from method findAllUsers

erastian committed:
Work branch: Logger expanded

- semantic error in mailer.service fix

mdziarkach commented on src/articles/articles.controller.ts:
```diff
@@ -91,7 +97,9 @@ export class ArticlesController extends BaseController implements IArticlesContr
 		} catch (e) {
-			return next(e);
+			return next(
+				new Exception(StatusCodes.NOT_FOUND, `Article "/${req.params.url}" not found`, ArticlesController.name),
```
we decided to roll it back

mdziarkach commented on src/services/errors/exception.filter.ts:
```diff
@@ -1,23 +1,54 @@
-	catch(err: Error | HTTPError, req: Request, res: Response, next: NextFunction): void {
-		if (err instanceof HTTPError) {
-			this.logger.error(`[${err.context}] Error ${err.statusCode}: ${err.message}`);
+	catch(err: Error | Exception | JsonWebTokenError, req: Request, res: Response, next: NextFunction): void {
```
We agreed that you would simplify this handler as much as possible and not complicate it. In all cases, the error has a name,code, and message. Why not show a uniform error?

erastian committed:
Work branch: Logger expanded

- moved catching error in getArticleByURL from controller to service layer
